make ObjectReference field ownership granular

alexzielenski opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug
/kind regression

#### What this PR does / why we need it:

This change changes the `structType` of `ObjectReference` from `atomic` (atomic added in v1.22 #100684). 

Something this PR is particularly concerned with is the transition path from users of 1.22 with atomic ownership, to having granular field management setting. An integration test was added to ensure the expected behavior.

Without this fix, users of SSA will find since the object is atomic, unspecified fields are cleared by the apiserver when applying changes to objects. This is a problem because there are often multiple agents writing to an `ObjectReference`. For example, in the case of `PersistentVolume`'s `claimRef`, the `uid` field and others might be populated by `kube-controller-manager` but the `name` or `namespace` fields would be specified by a user.

#### Which issue(s) this PR fixes:

https://github.com/fluxcd/flux2/issues/2827
issues with this behavior reported in [this comment](https://github.com/kubernetes/kubernetes/pull/100684#discussion_r888465096)

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Changes ownership semantics of PersistentVolume's spec.claimRef from `atomic` to `granular`.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
Note for Server-Side Apply Users: Following this change, existing atomic owners of PersistentVolumeClaim `spec.claimRef` fields will not own any of the nested fields. Nested fields will have whatever owner may be already specified in managed fields, or whoever claims the field first.
```

/assign @apelisse

apelisse commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -5635,7 +5635,7 @@ type ServiceProxyOptions struct {
 // For example, ServiceReferences for admission registration: https://github.com/kubernetes/api/blob/release-1.17/admissionregistration/v1/types.go#L533 .
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
-// +structType=atomic
+// +structType=granular
```
Do we want to keep it called out or should we just remove the structType marker altogether?

apelisse commented:
Thanks @alexzielenski, I'm curious what @liggitt will say :-)

aojea commented:
Kmews*

El jue., 9 jun. 2022 23:16, Alex Zielenski ***@***.***>
escribió:

> What type of PR is this?
>
> /kind bug
> /kind regression
> What this PR does / why we need it:
>
> This change changes the structType of ObjectReference from atomic (atomic
> added in v1.22 #100684
> <https://github.com/kubernetes/kubernetes/pull/100684>).
>
> Something this PR is particularly concerned with is the transition path
> from users of 1.22 with atomic ownership, to having granular field
> management setting. An integration test was added to ensure the expected
> behavior.
> Which issue(s) this PR fixes:
>
> fluxcd/flux2#2827 <https://github.com/fluxcd/flux2/issues/2827>
> Fixes #
> Special notes for your reviewer: Does this PR introduce a user-facing
> change?
>
> Changes ownership semantics of ObjectReference types from `atomic` to `granular`.
>
> Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals),
> usage docs, etc.:
>
> Note for Server-Side Apply Users: Following this change, existing atomic owners of `ObjectReference` fields will not own any of the nested fields. Nested fields will have whatever owner is specified in managed fields.
>
> /assign @apelisse <https://github.com/apelisse>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/kubernetes/kubernetes/pull/110495
> Commit Summary
>
>    - 1dcdb62
>    <https://github.com/kubernetes/kubernetes/pull/110495/commits/1dcdb62937c6420780c06d58c912c3feea35f0b0>
>    make ObjectReference field ownership granular
>
> File Changes
>
> (11 files <https://github.com/kubernetes/kubernetes/pull/110495/files>)
>
>    - *M* api/openapi-spec/swagger.json
>    <https://github.com/kubernetes/kubernetes/pull/110495/files#diff-ce7b64e72add3d44567b11ab1a4b1e6bfc9345a86b4a352283140e4974776c4d>
>    (2)
>    - *M* api/openapi-spec/v3/api__v1_openapi.json
>    <https://github.com/kubernetes/kubernetes/pull/110495/files#diff-32d8d3f65b972f670c1bfc3cb75b072ec9424642ebbaf328bdeccdc919c78c84>
>    (2)
>    - *M* api/openapi-spec/v3/apis__batch__v1_openapi.json
>    <https://github.com/kubernetes/kubernetes/pull/110495/files#diff-a9f8b0686411291b143dc526ce4ae478903dbfbe9c2d5831331404c5fa5d5f4f>
>    (2)
>    - *M* api/openapi-spec/v3/apis__discovery.k8s.io__v1_openapi.json
>    <https://github.com/kubernetes/kubernetes/pull/110495/files#diff-9804a76b3af29bddc0fda57eaa2df6c058bf224dfa3ea81cf8357653e8f983d5>
>    (2)
>    - *M* api/openapi-spec/v3/apis__events.k8s.io__v1_openapi.json
>    <https://github.com/kubernetes/kubernetes/pull/110495/files#diff-7af1765cc467c5fe08d1bd40b6bdba6795e198142607d123c348bad1200bc72d>
>    (2)
>    - *M* api/openapi-spec/v3/apis__storage.k8s.io__v1_openapi.json
>    <https://github.com/kubernetes/kubernetes/pull/110495/files#diff-940eff8ea5539eaef866e270d95a472c30b147f3efae5f377857c22fe82b7788>
>    (2)
>    - *M* pkg/generated/openapi/zz_generated.openapi.go
>    <https://github.com/kubernetes/kubernetes/pull/110495/files#diff-9f4c1466e676f9e733cae72d369ffd5ff37f446116c98562807cf3bfb872ae95>
>    (2)
>    - *M* staging/src/k8s.io/api/core/v1/generated.proto
>    <https://github.com/kubernetes/kubernetes/pull/110495/files#diff-22a65693efec8e1bbaef93a230cc438a588551f78f5df8e10a6a4d8cabc9df7e>
>    (2)
>    - *M* staging/src/k8s.io/api/core/v1/types.go
>    <https://github.com/kubernetes/kubernetes/pull/110495/files#diff-5ffbebb40a29f8e491f4ff1e07c618eb3edcc0d5f00a1a71445272d6767d7883>
>    (2)
>    - *M*
>    staging/src/k8s.io/client-go/applyconfigurations/internal/internal.go
>    <https://github.com/kubernetes/kubernetes/pull/110495/files#diff-d7743328e88b67741ee281243ae21b0b2911a65d13c3aad9938451d717eb3765>
>    (2)
>    - *M* test/integration/apiserver/apply/apply_test.go
>    <https://github.com/kubernetes/kubernetes/pull/110495/files#diff-b629323ef67f91837456deae39526a0e81aa0a5397465ff546e6845211afb13d>
>    (209)
>
> Patch Links:
>
>    - https://github.com/kubernetes/kubernetes/pull/110495.patch
>    - https://github.com/kubernetes/kubernetes/pull/110495.diff
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/110495>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABRGXIKVLUZ2RWUCSMQO2SLVOJNLTANCNFSM5YLNKKVQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

liggitt commented on staging/src/k8s.io/client-go/applyconfigurations/internal/internal.go:
```diff
@@ -5248,7 +5248,7 @@ var schemaYAML = typed.YAMLObject(`types:
       type:
         scalar: string
-    elementRelationship: atomic
+    elementRelationship: separable
```
what are the implications for client-go consumers with old versions of this?

liggitt commented on test/integration/apiserver/apply/apply_test.go:
```diff
@@ -3408,3 +3408,206 @@ func TestSubresourceField(t *testing.T) {
+		Post().
+		Param("fieldManager", "apply_test").
+		Resource("persistentvolumes").
+		Body(oldPersistentVolume).
```
I'm surprised this actually worked, trying to create an object with an existing uid set. Did you verify the persisted/returned object had the managedFields set you expected?

apelisse replied:
Yeah, I would expect all the fields set in the request to be owned by the POSTer here.

apelisse replied:
I think to be sure that you have the managedfields that you want, it'd be simpler to:
1. Create the object,
2. Send a patch (non-apply) that changes nothing in the object but the managedFields

alexzielenski replied:
Added separate update for managed fields. also added checks at the end to verify the managed fields were to be expected

liggitt commented on test/integration/apiserver/apply/apply_test.go:
```diff
@@ -3408,3 +3408,206 @@ func TestSubresourceField(t *testing.T) {
+	}
+	`)
+
+	// oldPersistentVolume, err := yamlutil.ToJSON(oldPersistentVolume)
+	// if err != nil {
+	// 	t.Fatalf("Failed marshal original yaml: %v", err)
+	// }
+
+	// originalPV, err = yamlutil.ToJSON(originalPV)
+	// if err != nil {
+	// 	t.Fatalf("Failed marshal yaml: %v", err)
+	// }
```
drop commented lines?

liggitt commented:
I see the case for this being a bug in PV, but most of the other uses still seem like atomic is correct.

Can we drop the structType and keep the other uses atomic by decorating the field at point of use?

apelisse commented:
> I see the case for this being a bug in PV, but most of the other uses still seem like atomic is correct.

Wouldn't you agree that we can't generally assume the UID  will be set by the same controller/actor as the gvk?

fedebongio commented:
/triage accepted

apelisse commented:
> > Wouldn't you agree that we can't generally assume the UID will be set by the same controller/actor as the gvk?
> 
> Yeah, probably not generally... that's an argument in favor of dropping the structType decoration and adding the field-level decorations to indicate it is atomic in most of the places we use the struct (if that even works)

Yeah, that sounds reasonable. I want to note that non-atomic gives much more control than atomic, there's no easy way to fix a zealous atomic marker with server-side apply, if you want to set the whole non-atomic struct, you can still set uid to null or something to unset it.

alexzielenski commented:
> > > I see the case for this being a bug in PV, but most of the other uses still seem like atomic is correct.
> > 
> > 
> > Wouldn't you agree that we can't generally assume the UID will be set by the same controller/actor as the gvk?
> 
> Yeah, probably not generally... that's an argument in favor of dropping the structType decoration and adding the field-level decorations to indicate it is atomic in most of the places we use the struct (if that even works)

I've attempted to add field-level decorators onto `PersistentVolume`'s `claimRef` which do have an effect on the resultant openapi; but they do not seem to override the type-level `atomic` specification.

Another alternative to making `ObjectReference` granular would be to make a `PersistentVolumeClaimReference` type for this use case? I am unsure how to assess which is more appropriate/less disruptive.

apelisse commented:
Let's look at occurences of ObjectReference and maybe see how they can be used.

alexzielenski commented:
re: @liggitt's comments

> Can we drop the structType and keep the other uses atomic by decorating the field at point of use?

This is the current state of the PR. I have removed the `structType` annotation from `ObjectReference` and added `field-level` annotations at point of use. However, done this way, the field-level annotations are not respected. I'm not sure what the proper way to do this would be using the current tools.

> Alternatively, can we keep structType as atomic and override the PV use to be granular by decorating the field... because this is a core type, changing structType will have implications for all out-of-tree consumers that compile this into their APIs as well

I've attempted this and unfortunately the k8s apiserver does not give precedence to the field-level annotation on the PV's field over the `ObjectReference`'s type-level annotation.

liggitt commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -337,6 +337,7 @@ type PersistentVolumeSpec struct {
 	// claim.VolumeName is the authoritative bind between PV and PVC.
 	// More info: https://kubernetes.io/docs/concepts/storage/persistent-volumes#binding
 	// +optional
+	// +structType=granular
```
I'm not sure our SSA code honors this annotation on a field... based on your experiments, it seems like it doesn't?

alexzielenski replied:
Correct, SSA does not honor it. I only pushed the code to share the diff/experiement while we decide how to move forward. To me our choices seem to be to either:

1.) Drop `atomic` from ObjectReference, make SSA honor field-level annotations and attach `atomic` as a field-level reference where necessary.
2.) Give `PersistentVolume` its own `ObjectReference` type

liggitt replied:
1 seems best, though its unfortunate it requires SSA changes

2 and 3 leave ObjectReference in a non-ideal state for other out of tree consumers

3 is too disruptive for a patch backport (breaks go compatibility of API types)

alexzielenski replied:
I am in favor of #1 in two parts:

a. Drop `atomic` from `ObjectReference`
b. Update `SSA` to allow field-level annotations of `structType`

apelisse replied:
Agreed, we need to stop the bleeding with a., we can do cosmetic changes with b. as needed later.

liggitt replied:
if we change a and backport it, we've also modified all the other API uses of this type... have we considered whether *those* changes will have negative consequences?

apelisse replied:
I'm not aware of anything client-wise (especially in client-go) using the structType to change the generation of anything. I'm not aware of anything client-side that uses this specifically to make any determination outside possibly of calling the Update/Apply function of the FieldManager. That includes kubectl who never uses this to do anything relative to Strategic-Merge-Patch or anything else.

The FieldManager has some logic to deal with such changes in schema (we have a test in the smd repo for these, please we now have a test in this PR to verify that it works with actual types). The main consequence of this change is that we'll have a dangling ownership of the struct which isn't going to be honored. It'll probably stay "forever" without expected impact.

alexzielenski committed:
update kube-openapi

alexzielenski committed:
mark persistentvolume's claimRef as granular

alexzielenski committed:
add regression test of formerly atomic claimRef

alexzielenski commented:
PR has been updated to use new SSA feature for a field-level specification to override a type-level `structType`. Now `PersistentVolumeSpec.claimRef` uses a granular `ObjectReference`, despite `ObjectReference` being marked as `atomic`.

cc @liggitt @apelisse

liggitt commented:
/lgtm
/approve

thanks for going the extra mile for compatibility here, and for the test coverage

liggitt commented:
/milestone v1.25

adding to the milestone, since this is a bugfix for a regression in behavior that we're likely to try to backport

janetkuo commented:
Is this ever backported to older versions? @alexzielenski

alexzielenski commented:
I don't think so, and I'm not sure that we would pick now to versions earlier than 1.25 when it was introduced.
